File: library/src/main/java/kim/jeonghyeon/androidlibrary/ui/binder/recyclerview/BaseRecyclerViewHolder.java
package kim.jeonghyeon.androidlibrary.ui.binder.recyclerview;

import androidx.annotation.NonNull;
import androidx.databinding.ViewDataBinding;
import androidx.recyclerview.widget.RecyclerView;


/**
 * Created by hyun.kim on 27/12/17.
 * <p>
 * refer to {{@link BaseRecyclerViewAdapter}}
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public abstract class BaseRecyclerViewHolder<VM> extends RecyclerView.ViewHolder {
    @NonNull
    private final ViewDataBinding mBinding;
    private VM viewModel;

    public BaseRecyclerViewHolder(ViewDataBinding binding) {
        super(binding.getRoot());
        mBinding = binding;

    }

    public void bind(VM viewModel) {
        this.viewModel = viewModel;
        mBinding.setVariable(getViewModelId(), viewModel);
        mBinding.executePendingBindings();
    }

    public VM getViewModel() {
        return viewModel;
    }

    /**
     * if viewModel name is not "model", override this method to set the other view model name
     *
     * @return viewModel Id
     */
    protected abstract int getViewModelId();
}

File: kotlin-simple-architecture/src/main/java/kim/jeonghyeon/androidlibrary/ui/binder/recyclerview/BaseRecyclerViewHolder.kt
package kim.jeonghyeon.androidlibrary.ui.binder.recyclerview

import androidx.databinding.ViewDataBinding
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.RecyclerView

/**
 * Created by hyun.kim on 27/12/17.
 *
 *
 * refer to {[BaseRecyclerViewAdapter]}
 */
abstract class BaseRecyclerViewHolder<VM : Any>(
    private val binding: ViewDataBinding,
    private val lifecycleOwner: LifecycleOwner?
) : RecyclerView.ViewHolder(binding.root) {
    lateinit var viewModel: VM

    fun bind(viewModel: VM) {
        this.viewModel = viewModel
        binding.setVariable(viewModelId, viewModel)
        binding.lifecycleOwner = lifecycleOwner
        binding.executePendingBindings()
    }

    /**
     * if viewModel name is not "model", override this method to set the other view model name
     *
     * @return viewModel Id
     */
    protected abstract val viewModelId: Int

}